### PR TITLE
[RAPTOR-3237] Adding validation to drum push

### PR DESCRIPTION
**Pull request by 4dahalibut**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

### CODE REVIEW

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 1]
```diff
@@ -448,8 +460,9 @@ def _prepare_docker_command(self, options, run_mode):
         docker_cmd = "docker run --rm --interactive  --user $(id -u):$(id -g) "
         docker_cmd_args = " -v {}:{}".format(options.code_dir, in_docker_model)
 
-        in_docker_cmd_list = sys.argv
+        in_docker_cmd_list = raw_arguments or sys.argv
```
Comment (yakov-g) [inline custom_model_runner/datarobot_drum/drum/drum.py:463]: Take sys.argv from here to the upper level ...

File: custom_model_runner/datarobot_drum/drum/push.py [hunk 2]
```diff
@@ -164,6 +165,55 @@ def print_model_started_dialogue(new_model_id):
...
+def validate_inference(config, options):
+    path = Path(config["validation"]["input"])
+    if not os.path.isabs(path):
+        path = Path(options.code_dir).joinpath(path)
+
+    options.input = path
+    options.output = "/dev/null"
+    options.negative_class_label = None
+    options.positive_class_label = None
+    raw_args_for_docker = "drum validate --input {input} -cd {code_dir}".format(
```
Comment (yakov-g) [inline custom_model_runner/datarobot_drum/drum/push.py:201]: How binary classification is validated?

File: custom_model_runner/datarobot_drum/drum/runtime.py [hunk 3]
```diff
@@ -39,8 +39,10 @@ def __exit__(self, exc_type, exc_value, exc_traceback):
         exc_msg = "".join(exc_msg_lines)
 
         run_mode = RunMode(self.options.subparser_name)
-        if self.options.show_stacktrace or (
-            run_mode == RunMode.SERVER and self.options.with_error_server
+        if (
+            not hasattr(self.options, "show_stacktrace")
```
[the remark this answers is not in the document]
Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/runtime.py:43]: Whoever added show_stacktrace didn't add it to everything, I don't want it in fit or push at all, so I think this fine and is forward looking.

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 1 above)
Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/drum.py:463]: I'm sending in raw_arguments = None by default so that wouldn't work.

Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/drum.py]: exactly set `raw_arguments = sys.argv` for default.
Anyhow, finish with docker implementation before

File: custom_model_runner/datarobot_drum/drum/push.py [hunk 4]
```diff
@@ -164,6 +165,55 @@ def print_model_started_dialogue(new_model_id):
...
+
+    return options, RunMode.FIT, raw_args_for_docker
+
+
+def validate_inference(config, options):
+    path = Path(config["validation"]["input"])
+    if not os.path.isabs(path):
+        path = Path(options.code_dir).joinpath(path)
+
+    options.input = path
```
Comment (yakov-g) [inline custom_model_runner/datarobot_drum/drum/push.py:197]: why do you work with original options, parsed for `push`;
Why not to create a new ones, which will hold only score/fit required options?

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 5]
```diff
@@ -224,37 +225,49 @@ def run(self):
...
+            print(
+                "Your model was successfully validated locally! Now we will add it into DataRobot"
+            )
             drum_push(self.options)
         else:
             error_message = "{} mode is not implemented".format(self.run_mode)
             print(error_message)
             raise DrumCommonException(error_message)
 
+    def run_fit(self):
```
Comment (liororama) [inline custom_model_runner/datarobot_drum/drum/drum.py:251]: Name of method is confusing.. maybe run_fit_validation?

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 6]
```diff
@@ -224,37 +225,49 @@ def run(self):
...
+        remove_temp_output = None
+        if not self.options.output:
+            self.options.output = mkdtemp()
+            remove_temp_output = self.options.output
+        self._run_fit_and_predictions_pipelines_in_mlpiper()
+        if self.options.output or not self.options.skip_predict:
+            create_custom_inference_model_folder(self.options.code_dir, self.options.output)
+        if not self.options.skip_predict:
+            self.run_test_predict()
+        if remove_temp_output:
```
Comment (liororama) [inline custom_model_runner/datarobot_drum/drum/drum.py:261]: Maybe I am missing something.. (probably).. but why different outputs in this case?
Validation is done in both cases.. yet the message is different?

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 5 above)
Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/drum.py:251]: sure

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 6 above)
Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/drum.py:261]: If there is no output argument supplied, we assumethey're using fit for validation and give them a big success message.

Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/drum.py:261]: If there is an output we assume they want to use drum to produce a new model.

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 5 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/drum.py:251]: Once you are going to change this, please add `_` to the internal methods.

Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/drum.py:251]: Sorry, this helper method is being used any time drum is running fit, it fits with the overall schema of naming I think

File: custom_model_runner/datarobot_drum/drum/push.py (hunk 2 above)
Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/push.py]: The predictions of the outputs are never checked, the data is just fed in.

**Comment by datarobot-jarvis** [comment]
:dollar: Total cost of executing tests in Jenkins: $**3.14** USD.

[What is this all about?](http://bit.ly/2BahK4V)

**Pull request merged**